lib/navigation/preprocess_encoders.cpp

```diff
@@ -2,55 +2,104 @@
+std::optional<EncodersPreprocessor::Statistics> EncodersPreprocessor::getStatistics(
+  const core::EncoderData &encoder_data) const
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - get q1, median, q3 of encoder array
-  - define upper & lower bounds as (-)1.5*inter-quatrile range
-  - if any datapoint outwith this, set
-  num_outliers_per_encoder_[i] += 1
-  -set outlier points as median
-
-  -also has to be able to handle 1 unreliable sensor
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
-  return {0, 0, 0, 0};
+  if (num_reliable_encoders_ < core::kNumEncoders - 1) {
+    logger_.log(core::LogLevel::kFatal,
+                "Unsuitable number of reliable encoders (%d of %d)",
+                num_reliable_encoders_,
+                core::kNumEncoders);
+    return std::nullopt;
+  }
+  if (num_reliable_encoders_ == core::kNumEncoders) {
+    auto reliable_data                    = encoder_data;
+    const Quartile quartiles              = getQuartiles(reliable_data);
+    const core::Float interquartile_range = quartiles.third_quartile - quartiles.first_quartile;
+    return {{.median      = quartiles.median,
+             .upper_bound = quartiles.median + 1.5F * interquartile_range,
+             .lower_bound = quartiles.median - 1.5F * interquartile_range}};
+  }
+  std::array<std::uint32_t, core::kNumEncoders - 1> reliable_data;
+  std::size_t j = 0;
+  for (std::size_t i = 0; i < encoder_data.size(); ++i) {
+    if (are_encoders_reliable_.at(i)) {
+      reliable_data.at(j) = encoder_data.at(i);
+      ++j;
+    }
+  }
+  const Quartile quartiles              = getQuartiles(reliable_data);
+  const core::Float interquartile_range = quartiles.third_quartile - quartiles.first_quartile;
+  return {{.median      = quartiles.median,
+           .upper_bound = quartiles.median + 1.2F * interquartile_range,
+           .lower_bound = quartiles.median - 1.2F * interquartile_range}};
 }
 
-void EncodersPreprocessor::checkReliable(const core::EncoderData &encoder_data)
+std::optional<core::EncoderData> EncodersPreprocessor::sanitise(
+  const core::EncoderData &encoder_data)
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - check how many times an individual encoder
-  has been an outlier in a row (outlier encoders)
-  -if num_outliers_per_encoder_[i] > n (tbd), mark encoder as unreliable
-  in reliable encoders.
-  -if number unreliable in reliable_encoders > 1, fail state
+  const auto encoder_statistics_optional = getStatistics(encoder_data);
+  if (!encoder_statistics_optional) {
+    logger_.log(core::LogLevel::kFatal, "Failed to obtain statistics for measurement");
+    return std::nullopt;
+  }
+  const auto encoder_statistics = *encoder_statistics_optional;
+  auto sanitised_data           = encoder_data;
+  for (std::size_t i = 0; i < sanitised_data.size(); ++i) {
+    if (sanitised_data.at(i) > encoder_statistics.upper_bound
```

Comment:
```
if(sanitised_data.at(i) > encoder_statistics.upper_bound && are_encoders_reliable.at(i) ||  sanitised_data.at(i) < encoder_statistics.lower_bound && are_encoders_reliable.at(i))
```
?
Currently if `encoders.at(i)` are unreliable, we keep adding to `num_consecutive_outliers_per_encoder.at(i)` which has the potential to get too big for a `uint16_t` to contain it on an extended run.

Reply:
Or break it up or apply this logic more appropriately.